docs/en/pact-reference.md

```diff
@@ -1277,9 +1277,200 @@ Declaring models shares the same syntax with modules:
+value references a concrete module that implements the `fungible-v2` interface,
+giving it the ability to call `fungible-v2` operations using those values.
+
+```
+;;; simplified DEX example with modref dynamic dispatch
+(defun swap
+  ( a-token:module{fungible-v2} a-amount:decimal a-account:string
+    b-token:module{fungible-v2} b-amount:decimal b-account:string
+  )
+  (with-read pair-accounts (format "{}:{}" [a-token b-token])
+    { 'pair-a-account := pair-a-account
+    , 'pair-b-account := pair-b-account
+    }
+    (a-token::transfer a-account pair-a-account a-amount)
+    (b-token::transfer pair-b-account b-account b-amount))
+)
+```
+
+To invoke the above function, the module names are directly referenced in code.
+
+```lisp
+
+(swap coin a-amount a-account
+      kbtc.ledger b-amount b-account)
+
+```
+
+
+The refmod values are "normal Pact values" that can be stored in the database,
+referenced in events and returned from functions.
+
+```
+;;; simplified DEX example with stored pair modrefs
+(defun swap
+  ( pair-symbol:string
+    a-amount:decimal a-account:string
+    b-amount:decimal b-account:string
+  )
+  (with-read pair-accounts pair-symbol
+    { 'pair-a-account := pair-a-account:string
+    , 'a-token := a-token:module{fungible-v2}
+    , 'pair-b-account := pair-b-account:string
+    , 'b-token := b-token:module{fungible-v2}
+    }
+    (a-token::transfer a-account pair-a-account a-amount)
+    (b-token::transfer pair-b-account b-account b-amount))
+)
+```
+
+
+### Modrefs and Polymorphism
+
+Modrefs provide polymorphism for use cases like the example above with an emphasis on
+interoperability. A modref is specified with one or more interfaces, allowing
+for values of that modref to reference modules that implement those
+interfaces.
+
+In the calling example above, the modref `a-token:module{fungible-v2}` accepts a
+reference to the Kadena `coin` KDA token module, because `coin` implements
+`fungible-v2`. Of course there is nothing special about `fungible-v2`: modrefs
+can specify any defined interface and accept any module that implements said
+interface.
+
+In this way, modrefs bring polymorphism to Pact. To programmers coming from
+polymorphic languages like Javascript, Java or Python, module references may
+look familiar. However, Pact is very much not object-oriented so there are
+important differences in how polymorphism works and what it is best used for.
+
+A key difference is that while modrefs allow polymorphism over modules, they
+don't offer polymorphism over data. Modules in Pact are not data in any
+practical sense: not only are they "code only", but they cannot be dynamically
+created within the Pact system.
+
+Meanwhile, data stored in the database is managed using schemas, which have no
+polymorphic features via modrefs or any other mechanism. This is why we suggest
+modrefs are more about "interoperation" instead of polymorphism per se, avoiding
```

Comment:
The length of this section about differences from standard object-oriented polymorphism makes me think that the word 'polymorphism' may be leading us somewhat astray. Perhaps 'generics' is a term in common use that more closely describes the purpose of module references? I think generics are understood to be about abstraction via an interface, and the ability to call functions / methods from that interface, without the same object-oriented baggage that seems to be associated with 'polymorphism'.

Reply:
I might be trying too hard. My goal was to say "see OO-ers, this isn't the polymorphism you're looking for" by saying that data objects in pact don't gain any polymorphism through modrefs that they're used to, because in OO objects function as both code and data.

Reply:
Sadly I can't find any literature that really nails down the differences between generics, parametric polymorphism, or ad-hoc polymorphism. Generics really have a language-specific meaning, where in C++ and Haskell it's codegen, but in Java it's adding type variables to interface declarations and leveraging OO polymorphism (subtyping) to get the job done.

Reply:
I went on that journey myself earlier today :laughing:. I think that it's OK to focus on what modrefs provide and not bother drawing a distinction between the abstraction / polymorphism provided by modrefs vs. standard object-oriented programming. I would expect someone reading the Pact reference would know they're not in OO world anymore.

That could certainly be useful to write up elsewhere, but for the reference it felt like an unexpected digression.

Reply:
I guess Rust decided to sidestep the whole issue with "traits", so we have yet another word in the parlance to deal with!
https://doc.rust-lang.org/book/ch10-02-traits.html

Reply:
Not sure how much we want to add references to more familiar languages, but for the javascript community, a parallel can be drawn between `types` of typescript (which @alber70g probably can explain better than I can)

Reply:
I reworked the section, please take a look! Thanks @thomashoneyman for the Rust example and @EnoF for the typescript, I mentioned both :)